It is totally the `np.nan`. The reason is that it is inside a structured array.


```
In [1]: import numpy as np                                                    

In [2]: a = np.zeros(1, dtype='f8,f8')                                        

In [3]: a                                                                     
Out[3]: array([(0., 0.)], dtype=[('f0', '<f8'), ('f1', '<f8')])

In [4]: a[0] = (np.nan, 0)                                                    

In [5]: from numpy.testing import assert_array_equal                          

In [6]: assert_array_equal(a, a)                                              
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-6-96868a4fad08> in <module>
----> 1 assert_array_equal(a, a)

/home/mark/git/numpy/numpy/testing/_private/utils.py in assert_array_equal(x, y, err_msg, verbose)
    871     __tracebackhide__ = True  # Hide traceback for py.test
    872     assert_array_compare(operator.__eq__, x, y, err_msg=err_msg,
--> 873                          verbose=verbose, header='Arrays are not equal')
    874 
    875 

/home/mark/git/numpy/numpy/testing/_private/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    795                                 verbose=verbose, header=header,
    796                                 names=('x', 'y'), precision=precision)
--> 797             raise AssertionError(msg)
    798     except ValueError:
    799         import traceback

AssertionError: 
Arrays are not equal

(mismatch 100.0%)
 x: array([(nan, 0.)], dtype=[('f0', '<f8'), ('f1', '<f8')])
 y: array([(nan, 0.)], dtype=[('f0', '<f8'), ('f1', '<f8')])
```

--- a/numpy/core/tests/test_records.py
+++ b/numpy/core/tests/test_records.py
@@ -332,15 +332,14 @@ class TestPathUsage(object):
     def test_tofile_fromfile(self):
         with temppath(suffix='.bin') as path:
             path = Path(path)
-            a = np.empty(10, dtype='f8,i4,a5')
+            np.random.seed(123)
+            a = np.random.rand(10).astype('f8,i4,a5')
             a[5] = (0.5,10,'abcde')
-            a.newbyteorder('<')
             with path.open("wb") as fd:
                 a.tofile(fd)
             x = np.core.records.fromfile(path,
                                          formats='f8,i4,a5',
-                                         shape=10,
-                                         byteorder='<')
+                                         shape=10)
             assert_array_equal(x, a)
 
 

